pymatgen/io/openmm/inputs.py

```diff
@@ -0,0 +1,301 @@
+import io
+import os
+from pathlib import Path
+import pathlib
+import re
+import shutil
+import tempfile
+import warnings
+from string import Template
+from typing import Union, Optional, Dict, List
+
+import numpy as np
+from monty.json import MSONable
+from monty.dev import deprecated
+
+import pymatgen.core
+from pymatgen.io.core import InputFile, InputSet, InputGenerator
+from pymatgen.io.packmol import PackmolBoxGen
+from pymatgen.io.babel import BabelMolAdaptor
+from pymatgen.io.xyz import XYZ
+
+import openff
+from openff.toolkit.typing.engines import smirnoff
+
+import openmm
+from openmm.app import Simulation, PDBFile, Topology
+from openmm import XmlSerializer, System, Integrator, State
+from openmm.unit import *
+
+from openbabel import pybel
+
+import parmed
+
+__author__ = "Orion Cohen"
+__version__ = "1.0"
+__maintainer__ = "Orion Cohen"
+__email__ = "orion@lbl.gov"
+__date__ = "Nov 2021"
+
+
+class TopologyInput(InputFile):
+    def __init__(self, topology: Topology, positions: Union[np.ndarray, List]):
+        self.topology = topology
+        self.positions = positions
+
+    def get_string(self) -> str:
+        with io.StringIO() as s:
+            PDBFile.writeFile(self.topology, self.positions)
+            s.seek(0)
+            pdb = s.read()
+        return pdb
+
+    @classmethod
+    def from_string(cls, contents: str):
+        with io.StringIO(contents) as s:
+            pdb = PDBFile(s)
+            topology = pdb.getTopology()
+            positions = pdb.getPositions(asNumpy=True)
+        return TopologyInput(topology, positions)
+
+
+class SystemInput(InputFile):
+    def __init__(self, system: System):
+        self.system = system
+
+    def get_string(self) -> str:
+        return XmlSerializer.serialize(self.system)
+
+    @classmethod
+    def from_string(cls, contents: str):
+        return SystemInput(XmlSerializer.deserialize(contents))
+
+
+class IntegratorInput(InputFile):
+    def __init__(self, integrator: Integrator):
+        self.integrator = integrator
+
+    def get_string(self) -> str:
+        return XmlSerializer.serialize(self.integrator)
+
+    @classmethod
+    def from_string(cls, contents: str):
+        return IntegratorInput(XmlSerializer.deserialize(contents))
+
+
+class StateInput(InputFile):
+    def __init__(self, state: State):
+        self.state = state
+
+    def get_string(self) -> str:
+        return XmlSerializer.serialize(self.state)
+
+    @classmethod
+    def from_string(cls, contents: str):
+        return StateInput(XmlSerializer.deserialize(contents))
+
+
+class OpenMMSet(InputSet):
+
+    # TODO: if there is an optional file should it be missing or should the value be none?
+    @classmethod
+    def from_directory(
```

Comment:
I assume you're talking about `state.xml`? If I read your current code right, it says "look for a file named `state.xml`, but that file is not required". I think the way you have it is OK. Are you asking whether the `state_file` kwarg should be `None` for an optional file?

Reply:
You interpreted my comment correctly. I am wondering if setting the `state_file` kwarg to `None` is more appropriate. It would look something like:

```python
    def from_directory(
        cls,
        directory: Union[str, Path],
        topology_file: str = "topology.pdb",
        system_file: str = "system.xml",
        integrator_file: str = "integrator.xml",
        state_file: Optional[str] = None,
    ):
        topology = TopologyInput.from_file(topology_file)
        system = SystemInput.from_file(system_file)
        integrator = IntegratorInput.from_file(integrator_file)
        openmm_set = OpenMMSet(
            topology=topology,
            system=system,
            integrator=integrator,
        )
        if state_file:
            state = StateInput.from_file(state_file)
            openmm_set["state"] = state
        return openmm_set
```